### PR TITLE
Report links in Word and other programs in braille.

### LINKED ISSUE #6780: Links in Word aren't displayed with the lnk indicator in braille
STR:
1. Open Word, tested with Word 2016.
2. Type "This is something to click on."
3. Select the word "click", right-click on it, choose Hyperlink... and make it a link.

Speech reads:
This is something to link click out of link on.
Braille shows:
This is something to click on.
Expected:
Braille somehow shows the link, e.g. with the "lnk" indicator.

### PR DESCRIPTION
**Pull request by dkager**
To implement this, use a format field attributes cache similar to the one used for speech. This can later be extended to cache other things.

It appears to work correctly, however there is one corner case:
- Open Word 2016.
- Type "Please clickhere." (Note: no space between click and here.)
- Select "here" and make it a link.
- Cursor over to the "h" and start arrowing to the left.
- The "lnk" text, i.e. the start indicator starts moving to the left too!

This looks like a Word-specific problem. When you add a space between "click" and "here" everything works as expected.

Fixes #6780

### CODE REVIEW

**Comment by leonardder** [comment]
> It appears to work correctly, however there is one corner case:

This corner case also shows up in speech, but there, clickhere is always considered to be a link e.g. the link indicator is always spoken before the word click). When copying the exact same text snippet to libre office, everything works as expected indeed.

**Review by jcsteh — CHANGES_REQUESTED** [review]
Thanks for your work.

The general approach looks good. However, I think we don't need something as complex as TextInfoRegionState with the updateObj behaviour, etc. I realise we do this for speech, but that's only because of say all and the fact that it queues several lines ahead, which required us to delay the cache update; see bd799f43. In contrast, we can't currently "queue ahead" for braille. I guess it's not inconceivable that we might one day support fetching ahead for braille continuous reading, but I think that's going to require a lot of other refactor and is probably best considered when we get there. For now, I think simpler is better.

So, you could either:

1. Store the format attribs cache dict directly on the NVDAObject.
2. Keep the TextInfoRegionState object, but it would only hold the format attribs cache. It doesn't need to track the object, support copying, updateObj, etc.

Either way, it'd be good to have the NVDAObject attribute name prefixed with "braille" so it's clear that this is for braille. "_textInfoRegionState" might not immediately suggest braille to the uninitiated.

**Comment by dkager** [comment]
At first glance I would go for the second option. Dumping more things onto NVDAObject seems a bad idea because most of these objects will not need this cache.

**Comment by dkager** [comment]
Never mind. TextInfoRegionState also updates NVDAObject, so that is even more confusing. I think the first approach will make much cleaner code.

**Comment by dkager** [comment]
@jcsteh Thanks for the review. Is this new implementation closer to what you had in mind?

File: source/braille.py [hunk 1]
```diff
@@ -792,7 +792,7 @@ def getControlFieldBraille(info, field, ancestors, reportStart, formatConfig):
 		return (_("%s end") %
 			getBrailleTextForProperties(role=role))
 
-def getFormatFieldBraille(field, isAtStart, formatConfig):
+def getFormatFieldBraille(field, isAtStart, formatConfig, fieldCache=None):
```
Comment (leonardder) [inline source/braille.py]: May be add a doc string here?

File: source/braille.py [hunk 2]
```diff
@@ -808,6 +808,14 @@ def getFormatFieldBraille(field, isAtStart, formatConfig):
 				# Translators: Displayed in braille for a heading with a level.
 				# %s is replaced with the level.
 				textList.append(_("h%s")%headingLevel)
+	if formatConfig["reportLinks"]:
+		link=field.get("link")
+		oldLink=fieldCache.get("link") if fieldCache is not None else None
```
Comment (leonardder) [inline source/braille.py]: Would it be bad coding style to provide an empty dictionary as default for fieldCache? That would allow you to get rid of the "if fieldCache is not None else None" part

File: source/braille.py [hunk 3]
```diff
@@ -878,10 +886,15 @@ def _addFieldText(self, text, contentPos):
 		self.rawTextTypeforms.extend((louis.plain_text,) * textLen)
 		self._rawToContentPos.extend((contentPos,) * textLen)
 
-	def _addTextWithFields(self, info, formatConfig, isSelection=False):
```
Comment (leonardder) [inline source/braille.py]: Another doc string missing which might be useful somehow, but I'm overly nagging now.

File: source/braille.py [hunk 4]
```diff
@@ -878,10 +886,15 @@ def _addFieldText(self, text, contentPos):
 		self.rawTextTypeforms.extend((louis.plain_text,) * textLen)
 		self._rawToContentPos.extend((contentPos,) * textLen)
 
-	def _addTextWithFields(self, info, formatConfig, isSelection=False):
+	def _addTextWithFields(self, info, formatConfig, isSelection=False, useCache=True):
 		shouldMoveCursorToFirstContent = not isSelection and self.cursorPos is not None
 		ctrlFields = []
 		typeform = louis.plain_text
+		if useCache:
```
Comment (leonardder) [inline source/braille.py]: How about checking this in line? That would collapse four lines into one.
```
			formatFieldAttributesCache = getattr(info.obj, "_brailleFormatFieldAttributesCache", {}) if useCache else None
```

Comment (dkager) [inline source/braille.py]: Using an empty dictionary is semantically a bit different from having no cache I guess. But if we find that acceptable, then we don't even have to check for `useCache` here.
However, note that I stole this code from speech.py, and I think it's good to keep those similar in code style.

File: source/braille.py (hunk 1 above)
Comment (dkager) [inline source/braille.py]: Any suggestions for the `isAtStart` parameter doc? You have been working with that recently I think.

Comment (jcsteh) [inline source/braille.py]: isAtStart: True if this FormatField precedes any text in the line/paragraph. This is useful to restrict display of information which should only appear at the start of the line/paragraph; e.g. the line number or line prefix (list bullet/number).

File: source/braille.py (hunk 2 above)
Comment (jcsteh) [inline source/braille.py]: 1. This would make it impossible to tell the difference between an empty cache and no cache. The distinction is important because it determines whether you update the cache at the end.
2. And if you do update the cache at the end regardless, the empty dictionary you specified as a default would no longer be empty for future runs. That's a common Python gotcha. Consider the following code:

> ```
> def f(l=[]):
>    l.append("test")
>    print l
>
> f()
> f()
> ```

Here's the output:

> ~~~
> ['test']
> ['test', 'test']
> ```

In short, you absolutely shouldn't do this. :)

**Review by jcsteh — COMMENTED** [review]
Looks good. Thanks.

Do you see any case where we'd ever set useCache to False? For speech, we do this when reporting formatting with NVDA+f. However, for braille, this would only report some of the available info, since some of it is communicated via other means (typeforms, etc.). If there's no case for it, I'd just drop the argument.

**Comment by dkager** [comment]
I don't see a case for not caching. If there ever turns out to be one it is easy enough to add this parameter back.

File: source/braille.py [hunk 5]
```diff
@@ -792,7 +792,19 @@ def getControlFieldBraille(info, field, ancestors, reportStart, formatConfig):
 		return (_("%s end") %
 			getBrailleTextForProperties(role=role))
 
-def getFormatFieldBraille(field, isAtStart, formatConfig):
+def getFormatFieldBraille(field, fieldCache, isAtStart, formatConfig):
```
Comment (dkager) [inline source/braille.py]: This breaks backwards compatibility because a new parameter is added in the second position. However, this looked more logical than adding it to the end, because `field` and `fieldCache` are closely related.

File: source/braille.py [hunk 6]
```diff
@@ -808,6 +820,13 @@ def getFormatFieldBraille(field, isAtStart, formatConfig):
 				# Translators: Displayed in braille for a heading with a level.
 				# %s is replaced with the level.
 				textList.append(_("h%s")%headingLevel)
+	if formatConfig["reportLinks"]:
+		link=field.get("link")
+		oldLink=fieldCache.get("link")
+		if link and link != oldLink:
+			textList.append(roleLabels[controlTypes.ROLE_LINK])
+	fieldCache.clear()
```
Comment (dkager) [inline source/braille.py]: @jcsteh How is this for Python gotchas?

File: source/braille.py [hunk 7]
```diff
@@ -959,6 +979,7 @@ def _addTextWithFields(self, info, formatConfig, isSelection=False):
 			# We only render fields that aren't at the start of their nodes for the first part of the reading unit.
 			# Otherwise, we'll render fields that have already been rendered.
 			self._skipFieldsNotAtStartOfNode = True
+		info.obj._brailleFormatFieldAttributesCache = formatFieldAttributesCache
```
Comment (dkager) [inline source/braille.py]: My concern is that this will clutter many NVDAObjects with empty format field caches. But I didn't see a better way. Suggestions welcome!

**Review by dkager — COMMENTED** [review]
This new version hasn't been thoroughly tested yet. I first want to make sure my design makes sense.

File: source/braille.py (hunk 6 above)
Comment (jcsteh) [inline source/braille.py]: > @jcsteh How is this for Python gotchas?

All good. The empty dict is created inside the function, so it's local to each run of that function. The gotcha only applies for keyword arguments. Think of keyword argument defaults as being at global scope, even though you can't access them outside of the function.

File: source/braille.py (hunk 5 above)
Comment (jcsteh) [inline source/braille.py]: > This breaks backwards compatibility because a new parameter is added in the second position.

I think this is okay. We don't call this function anywhere else and I couldn't find any matches grepping through all the add-ons hosted on the official repository. (This isn't full coverage, but I think it's highly unlikely that anyone ever calls this outside of braille itself.)

File: source/braille.py (hunk 7 above)
Comment (jcsteh) [inline source/braille.py]: > My concern is that this will clutter many NVDAObjects with empty format field caches.

Only objects which were at some point the focus or navigator object, and no more clutter than `speakTextInfo` produces. I guess you could test whether the updated cache is empty and if so, ensure the NVDAObject attribute is removed. However, I don't think there's really any point.

**Comment by jcsteh** [comment]
> This new version hasn't been thoroughly tested yet. I first want to make sure my design makes sense.

This looks fine to me. Happy to approve once tested.

**Comment by dkager** [comment]
Tested and it works. One observation is that you can't activate a link by pressing cursor routing keys on the "lnk" control type text. You can route the cursor by pressing there, though. This is the same behavior as in browse mode.

I was wondering if we should allow the `fieldCache` parameter to be `None` after all. Or is it enough to set it to `{}` to indicate an empty cache? I am thinking in particular about cases where there is only one call to `getFormatFieldBraille`.

**Comment by jcsteh** [comment]
When would there only be one call to getFormatFieldBraille? An empty cache
will still get populated and checked. If we do want to support None, we'll
need to reinstate all the code that checks for it right down the chain.

**Comment by dkager** [comment]
> When would there only be one call to getFormatFieldBraille?

I don't know. If there is only one object being brailled? Even so, the empty cache is not a problem. (I was trying to find reasons why the code could fail, but am coming up empty.) Looks good to me now.

**Review by jcsteh — APPROVED** [review]
